> Does CI pass if you add build to the targets run by the CI workflow? The wolkflow is at: .github/workflows/ci.yaml

First try doesn't work:
`...`
`  CC        /home/runner/work/tock-on-titan/tock-on-titan/userspace/gpio_test/../../third_party/libtock-c/libtock/internal/alarm_internal.c`
`/bin/sh: 1: arm-none-eabi-gcc: not found`
`make[1]: *** [/home/runner/work/tock-on-titan/tock-on-titan/userspace/gpio_test/../../build/libtock/cortex-m0/alarm_internal.o] Error 127`
`...`

--- a/.github/workflows/ci.yaml
+++ b/.github/workflows/ci.yaml
@@ -41,15 +41,15 @@ jobs:
       - name: Make setup
         run: make -C "${GITHUB_WORKSPACE}" setup
 
-      # Install the bubblewrap sandboxing tool.
-      - name: Install Bubblewrap
-        run: sudo apt-get install bubblewrap
+      # Install packages our build system depends on.
+      - name: Install Packages
+        run: sudo apt-get install bubblewrap gcc-arm-none-eabi
 
-      # Runs the `check` and `localtests` make actions. We cannot build this
-      # repository in GitHub Actions as building requires non-publicly-available
-      # code signing tools, so this is the best we can do (at least for now --
-      # theoretically we could build up until the code signing step but the
-      # makefiles are not structured for that yet). I use -j2 because the VMs
-      # that GitHub Actions uses have 2 logical CPUs.
-      - name: Check and local tests
-        run: make -C "${GITHUB_WORKSPACE}" -j2 check localtests
+      # Runs the `build`, `check` and `localtests` make actions.  We
+      # cannot build the `build-signed` target as doing so requires
+      # non-publicly-available code signing tools, so this is the best
+      # we can do (at least for now).
+      # We use -j2 because the VMs that GitHub Actions uses have 2
+      # logical CPUs.
+      - name: Build, check and local tests
+        run: make -C "${GITHUB_WORKSPACE}" -j2 build check localtests

--- a/Makefile
+++ b/Makefile
@@ -12,6 +12,9 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+# Note that `all` does _not_ depend on `build-signed`.  Many consumers
+# of this project (especially new ones) won't be happy if they have to
+# struggle with signing from day one.
 .PHONY: all
 all: build
 
@@ -54,6 +57,9 @@ sandbox_setup:
 
 .PHONY: build
 build: $(addsuffix /build,$(BUILD_SUBDIRS))
+
+.PHONY: build-signed
+build-signed: $(addsuffix /build-signed,$(BUILD_SUBDIRS))
 
 .PHONY: check
 check: $(addsuffix /check,$(BUILD_SUBDIRS))

--- a/README.md
+++ b/README.md
@@ -31,11 +31,17 @@ Configure Rust
 make setup
 ```
 
-### Build all boards and apps
+### Build all boards and apps (unsigned)
 
 ```shell
-make
+make build
 ```
 
-Note that if one of TANGO_CODESIGNER{,_KEY} is not set, then signed artifacts
-will not be created.
+### Build 'signed' versions of all artifacts
+
+```shell
+make build-signed
+```
+
+The `build-signed` target requires `TANGO_CODESIGNER` and `TANGO_CODESIGNER_KEY`
+to be set. The codesigner and keys are not publicly available.

--- a/doc/build_system.md
+++ b/doc/build_system.md
@@ -16,6 +16,7 @@ for instance, `userspace/h1_tests/Build.mk` should implement the following
 actions:
 
 * `userspace/h1_tests/build`
+* `userspace/h1_tests/build-signed`
 * `userspace/h1_tests/check`
 * `userspace/h1_tests/devicetests`
 * `userspace/h1_tests/doc`
@@ -38,6 +39,7 @@ exceptions to this (documented below).
 Each Build.mk should support the following targets:
 
 * `$PATH/build`
+* `$PATH/build-signed`
 * `$PATH/check`
 * `$PATH/devicetests`
 * `$PATH/doc`

--- a/doc/make_targets.md
+++ b/doc/make_targets.md
@@ -6,7 +6,7 @@ Common Makefile Targets
 For consistency, we want to expose the same `make` targets in each Makefile. For
 simplicity of development and testing, we want `make` commands run at the root
 of the repository to operate throughout the project. For example, running `make
-build` in `tock-on-titan/` should build all targets in the repository.
+build` in `tock-on-titan/` should build all unsigned targets in the repository.
 
 This specification defines common target names for the `tock-on-titan`
 repository. It is meant to be a reference for users of the build system (e.g.
@@ -26,10 +26,11 @@ Unless otherwise noted, the following targets operate in the directory in which
 <action> ...`. All per-directory makefiles support these targets, even if they
 do nothing (e.g. `make devicetests` in a directory with no device tests is a
 no-op but will succeed):
-
 * `all` is an alias for `build` (following Makefile tradition).
-* `build` compiles the code. This can be a library, an application image, the
-  kernel image, or the full flash image (kernel + applications).
+* `build` compiles the code, yielding unsigned artifacts. This can be a library,
+  an application image, the kernel image, or the full flash image (kernel +
+  applications).
+* `build-signed` compiles the code (via `build`), and then signs all artifacts.
 * `check` runs `cargo check`-style tests.
 * `clean` removes all build artifacts we know how to remove. It applies to the
   repository, not just the subdirectory it is called in.

--- a/kernel/Build.mk
+++ b/kernel/Build.mk
@@ -16,6 +16,9 @@
 kernel/build: sandbox_setup
 	cd kernel && $(BWRAP) cargo build --release
 
+.PHONY: kernel/build-signed
+kernel/build-signed: kernel/build
+
 .PHONY: kernel/check
 kernel/check: sandbox_setup
 	cd kernel && $(BWRAP) cargo check --release

--- a/runner/Build.mk
+++ b/runner/Build.mk
@@ -15,12 +15,15 @@
 .PHONY: runner/build
 runner/build: build/cargo-host/release/runner
 
+.PHONY: runner/build-signed
+runner/build-signed: runner/build
+
 .PHONY: runner/check
 runner/check: sandbox_setup
 	cd runner && \
 		CARGO_TARGET_DIR="../build/cargo-host" $(BWRAP) cargo check
 
-.PHONY: kernel/clean
+.PHONY: runner/clean
 runner/clean:
 	rm -f runner/Cargo.lock
 

--- a/tools/Build.mk
+++ b/tools/Build.mk
@@ -16,6 +16,9 @@
 tools/build: cargo_version_check sandbox_setup
 	cd tools && $(BWRAP) cargo build --offline --release
 
+.PHONY: tools/build-signed
+tools/build-signed: tools/build
+
 .PHONY: tools/check
 tools/check: cargo_version_check sandbox_setup
 	cd tools && $(BWRAP) cargo check --offline --release

--- a/userspace/Build.mk
+++ b/userspace/Build.mk
@@ -40,6 +40,10 @@ TANGO_BOARD_FOR_TEST ?= golf2
 .PHONY: userspace/build
 userspace/build: $(addsuffix /build,$(BUILD_SUBDIRS))
 
+.PHONY: userspace/build-signed
+userspace/build-signed: userspace/build
+userspace/build-signed: $(addsuffix /build-signed,$(BUILD_SUBDIRS))
+
 .PHONY: userspace/check
 userspace/check: sandbox_setup
 	cd userspace && TOCK_KERNEL_VERSION=h1_tests $(BWRAP) cargo check --release
@@ -82,15 +86,15 @@ include $(addsuffix /Build.mk,$(BUILD_SUBDIRS))
 
 
 # ------------------------------------------------------------------------------
-# Macro to define a full_image target for a specific
+# Macro to define unsigned_image and full_image targets for a specific
 # board-and-app-and-tbf-file combination.
 # Arguments:
 # - $(BOARD)
 # - $(APP)
 # - $(TBF_FILE)
-define FULL_IMAGE_TARGET
+define IMAGE_TARGETS
 
-build/userspace/$(APP)/$(BOARD)/full_image: \
+build/userspace/$(APP)/$(BOARD)/unsigned_image: \
 		build/userspace/$(APP)/$(TBF_FILE) \
 		kernel/build
 	mkdir -p build/userspace/$(APP)/$(BOARD)/
@@ -101,16 +105,14 @@ build/userspace/$(APP)/$(BOARD)/full_image: \
 	arm-none-eabi-objcopy --update-section \
 		.apps=build/userspace/$(APP)/$(TBF_FILE) \
 		build/userspace/$(APP)/$(BOARD)/unsigned_image
-	if [ -n "${TANGO_CODESIGNER}" -a -n "${TANGO_CODESIGNER_KEY}" ]; then \
-		$(TANGO_CODESIGNER) --b --input build/userspace/$(APP)/$(BOARD)/unsigned_image \
-			--key=$(TANGO_CODESIGNER_KEY) \
-			--output=build/userspace/$(APP)/$(BOARD)/signed_image; \
-		cat $(TANGO_BOOTLOADER) build/userspace/$(APP)/$(BOARD)/signed_image \
-			> build/userspace/$(APP)/$(BOARD)/full_image; \
-	else \
-		echo "***** Can't create signed_image -- one of CODESIGNER{,_KEY} is empty "; \
-		echo "*****   -- hope that's OK."; \
-	fi
+
+build/userspace/$(APP)/$(BOARD)/full_image: \
+		build/userspace/$(APP)/$(BOARD)/unsigned_image
+	$(TANGO_CODESIGNER) --b --input build/userspace/$(APP)/$(BOARD)/unsigned_image \
+		--key=$(TANGO_CODESIGNER_KEY) \
+		--output=build/userspace/$(APP)/$(BOARD)/signed_image;
+	cat $(TANGO_BOOTLOADER) build/userspace/$(APP)/$(BOARD)/signed_image \
+		> build/userspace/$(APP)/$(BOARD)/full_image;
 
 endef
 
@@ -122,8 +124,16 @@ endef
 # - $(APP)
 define C_APP_BOARD_TARGETS
 
+# `foreach TBF_FILE` ensures that TBF_FILE has the expected value when
+# IMAGE_TARGETS is expanded.
+$(foreach TBF_FILE,cortex-m3/cortex-m3.tbf,$(eval $(IMAGE_TARGETS)))
+
 .PHONY: userspace/$(APP)/$(BOARD)/build
 userspace/$(APP)/$(BOARD)/build: \
+		build/userspace/$(APP)/$(BOARD)/unsigned_image
+
+.PHONY: userspace/$(APP)/$(BOARD)/build-signed
+userspace/$(APP)/$(BOARD)/build-signed: \
 		build/userspace/$(APP)/$(BOARD)/full_image
 
 .PHONY: userspace/$(APP)/$(BOARD)/check
@@ -153,10 +163,6 @@ userspace/$(APP)/$(BOARD)/run: \
 		stty -F /dev/ttyUltraConsole3 115200 -echo ; \
 		stty -F /dev/ttyUltraTarget2 115200 -icrnl ; \
 		build/cargo-host/release/runner'
-
-# `foreach` ensures that TBF_FILE has the expected value when
-# FULL_IMAGE_TARGET is expanded.
-$(foreach TBF_FILE,cortex-m3/cortex-m3.tbf,$(eval $(FULL_IMAGE_TARGET)))
 
 endef
 
@@ -188,13 +194,19 @@ $(foreach APP,$(C_APPS),$(eval $(C_APP_TARGETS)))
 # - $(APP)
 define C_APP_COMBINED_TARGET
 
-# Entry target for an individual app to build app for all boards
+# Entry target for an individual app to build unsigned app for all boards
 .PHONY: userspace/$(APP)/build
 userspace/$(APP)/build: \
 		$(if $(filter $(APP),$(C_APPS)),$(foreach BOARD,$(BOARDS),userspace/$(APP)/$(BOARD)/build)) \
 		$(foreach BOARD,$(BOARDS),$(if $(filter $(APP),$(C_APPS_$(BOARD))),userspace/$(APP)/$(BOARD)/build))
 
-# Entry target for an individual app to build app for all boards
+# Entry target for an individual app to build signed app for all boards
+.PHONY: userspace/$(APP)/build-signed
+userspace/$(APP)/build-signed: \
+		$(if $(filter $(APP),$(C_APPS)),$(foreach BOARD,$(BOARDS),userspace/$(APP)/$(BOARD)/build-signed)) \
+		$(foreach BOARD,$(BOARDS),$(if $(filter $(APP),$(C_APPS_$(BOARD))),userspace/$(APP)/$(BOARD)/build-signed))
+
+# Entry target for an individual app to run check for all boards
 .PHONY: userspace/$(APP)/check
 userspace/$(APP)/check: \
 		$(if $(filter $(APP),$(C_APPS)),$(foreach BOARD,$(BOARDS),userspace/$(APP)/$(BOARD)/check)) \
@@ -233,8 +245,16 @@ $(foreach APP,$(C_APPS) $(foreach BOARD,$(BOARDS),$(C_APPS_$(BOARD))),$(eval $(C
 # - $(APP)
 define RUST_APP_BOARD_TARGETS
 
+# `foreach TBF_FILE` ensures that TBF_FILE has the expected value when
+# IMAGE_TARGETS is expanded.
+$(foreach TBF_FILE,$(BOARD)/app.tbf,$(eval $(IMAGE_TARGETS)))
+
 .PHONY: userspace/$(APP)/$(BOARD)/build
 userspace/$(APP)/$(BOARD)/build: \
+		build/userspace/$(APP)/$(BOARD)/unsigned_image
+
+.PHONY: userspace/$(APP)/$(BOARD)/build-signed
+userspace/$(APP)/$(BOARD)/build-signed: \
 		build/userspace/$(APP)/$(BOARD)/full_image
 
 .PHONY: userspace/$(APP)/$(BOARD)/check
@@ -294,10 +314,6 @@ build/userspace/$(APP)/$(BOARD)/app.tbf: \
 		     false ; \
 		fi
 
-# `foreach` ensures that TBF_FILE has the expected value when
-# FULL_IMAGE_TARGET is expanded.
-$(foreach TBF_FILE,$(BOARD)/app.tbf,$(eval $(FULL_IMAGE_TARGET)))
-
 endef
 
 # ------------------------------------------------------------------------------
@@ -313,11 +329,17 @@ $(foreach BOARD,$(BOARDS),$(foreach APP,$(RUST_APPS) $(RUST_APPS_$(BOARD)),$(eva
 # - $(APP)
 define RUST_APP_TARGET
 
-# Entry target for an individual app to build app for all boards
+# Entry target for an individual app to build unsigned app for all boards
 .PHONY: userspace/$(APP)/build
 userspace/$(APP)/build: \
 		$(if $(filter $(APP),$(RUST_APPS)),$(foreach BOARD,$(BOARDS),userspace/$(APP)/$(BOARD)/build)) \
 		$(foreach BOARD,$(BOARDS),$(if $(filter $(APP),$(RUST_APPS_$(BOARD))),userspace/$(APP)/$(BOARD)/build))
+
+# Entry target for an individual app to build signed app for all boards
+.PHONY: userspace/$(APP)/build-signed
+userspace/$(APP)/build-signed: \
+		$(if $(filter $(APP),$(RUST_APPS)),$(foreach BOARD,$(BOARDS),userspace/$(APP)/$(BOARD)/build-signed)) \
+		$(foreach BOARD,$(BOARDS),$(if $(filter $(APP),$(RUST_APPS_$(BOARD))),userspace/$(APP)/$(BOARD)/build-signed))
 
 # Entry target for an individual app to run check for all boards
 .PHONY: userspace/$(APP)/check
@@ -357,8 +379,16 @@ $(foreach APP,$(RUST_APPS) $(foreach BOARD,$(BOARDS),$(RUST_APPS_$(BOARD))),$(ev
 # - $(APP)
 define RUST_TEST_BOARD_TARGETS
 
+# `foreach TBF_FILE` ensures that TBF_FILE has the expected value when
+# IMAGE_TARGETS is expanded.
+$(foreach TBF_FILE,$(BOARD)/app.tbf,$(eval $(IMAGE_TARGETS)))
+
 .PHONY: userspace/$(APP)/$(BOARD)/build
 userspace/$(APP)/$(BOARD)/build: \
+		build/userspace/$(APP)/$(BOARD)/unsigned_image
+
+.PHONY: userspace/$(APP)/$(BOARD)/build-signed
+userspace/$(APP)/$(BOARD)/build-signed: \
 		build/userspace/$(APP)/$(BOARD)/full_image
 
 .PHONY: userspace/$(APP)/$(BOARD)/check
@@ -426,10 +456,6 @@ build/userspace/$(APP)/$(BOARD)/app.tbf: \
 		     false ; \
 		fi
 
-# `foreach` ensures that TBF_FILE has the expected value when
-# FULL_IMAGE_TARGET is expanded.
-$(foreach TBF_FILE,$(BOARD)/app.tbf,$(eval $(FULL_IMAGE_TARGET)))
-
 endef
 
 # ------------------------------------------------------------------------------
@@ -445,11 +471,17 @@ $(foreach BOARD,$(BOARDS),$(foreach APP,$(RUST_TESTS) $(RUST_TESTS_$(BOARD)),$(e
 # - $(APP)
 define RUST_TEST_TARGET
 
-# Entry target for an individual app to build app for all boards
+# Entry target for an individual app to build unsigned app for all boards
 .PHONY: userspace/$(APP)/build
 userspace/$(APP)/build: \
 		$(if $(filter $(APP),$(RUST_TESTS)),$(foreach BOARD,$(BOARDS),userspace/$(APP)/$(BOARD)/build)) \
 		$(foreach BOARD,$(BOARDS),$(if $(filter $(APP),$(RUST_TESTS_$(BOARD))),userspace/$(APP)/$(BOARD)/build))
+
+# Entry target for an individual app to build signed app for all boards
+.PHONY: userspace/$(APP)/build-signed
+userspace/$(APP)/build-signed: \
+		$(if $(filter $(APP),$(RUST_TESTS)),$(foreach BOARD,$(BOARDS),userspace/$(APP)/$(BOARD)/build-signed)) \
+		$(foreach BOARD,$(BOARDS),$(if $(filter $(APP),$(RUST_TESTS_$(BOARD))),userspace/$(APP)/$(BOARD)/build-signed))
 
 # Entry target for an individual app to run check for all boards
 .PHONY: userspace/$(APP)/check

